src/rgw/driver/rados/rgw_bucket.cc

```diff
@@ -317,98 +317,216 @@ static void dump_index_check(map<RGWObjCategory, RGWStorageStats> existing_stats
+				   entries_window.cbegin(),
+				   entries_window.cend());
+	}
 
-  for (const auto& o : all_objs) {
-    const std::string& name = o.second;
-    if (meta_objs.find(name) == meta_objs.end()) {
-      objs_to_unlink.push_back(o.first);
-    }
+	// either way start over
+	entries_window.clear();
+	prev_entry_prefix.clear();
+      } else {
+	if (entry_prefix != prev_entry_prefix) {
+	  entries_to_unlink.insert(entries_to_unlink.end(),
+				   entries_window.cbegin(),
+				   entries_window.cend());
+	  entries_window.clear();
+	  prev_entry_prefix = entry_prefix;
+	}
+
+	// create an rgw_obj_index_key to store in window
+	rgw_obj_index_key obj_index_key;
+	obj_key.get_index_key(&obj_index_key);
+	entries_window.push_back(obj_index_key);
+      }
 
-    if (objs_to_unlink.size() > listing_max_entries) {
-      if (fix_index) {
-	// note: under rados this removes directly from rados index objects
-	int r = bucket->remove_objs_from_index(dpp, objs_to_unlink);
-	if (r < 0) {
-	  set_err_msg(err_msg, "ERROR: remove_obj_from_index() returned error: " +
-		      cpp_strerror(-r));
-	  return r;
+      // check if this is a good point for intermediate index clean-up
+      if (entries_to_unlink.size() >= listing_max_entries) {
+	dump_multipart_index_results(entries_to_unlink,
+				     flusher.get_formatter());
+	if (fix_index) {
+	  store->remove_objs_from_index(dpp, bucket->get_info(),
+					entries_to_unlink);
 	}
+	entries_to_unlink.clear();
       }
+    } // for
+  } while (is_truncated);
+
+  // any entries left over at the end can be unlinked
+  entries_to_unlink.insert(entries_to_unlink.end(),
+			   entries_window.cbegin(),
+			   entries_window.cend());
+  entries_window.clear();
 
-      dump_mulipart_index_results(objs_to_unlink, f);
-      flusher.flush();
-      objs_to_unlink.clear();
+  if (! entries_to_unlink.empty()) {
+    dump_multipart_index_results(entries_to_unlink,
+				 flusher.get_formatter());
+    if (fix_index) {
+      store->remove_objs_from_index(dpp, bucket->get_info(),
+				    entries_to_unlink);
     }
+    entries_to_unlink.clear();
   }
 
-  if (fix_index) {
-    // note: under rados this removes directly from rados index objects
-    int r = bucket->remove_objs_from_index(dpp, objs_to_unlink);
-    if (r < 0) {
-      set_err_msg(err_msg, "ERROR: remove_obj_from_index() returned error: " +
-              cpp_strerror(-r));
+  flusher.flush();
 
-      return r;
-    }
+  return 0;
+} // static ::check_bad_index_multipart
+
+
+/**
+ * Checks for damaged incomplete multipart uploads in a bucket
+ * index. Since all entries for a given multipart upload end up on the
+ * same shard (by design), we spawn a set of co-routines, each one
+ * working shard by shard until all work is complete.
+ */
+int RGWBucket::check_bad_index_multipart(rgw::sal::RadosStore* const rados_store,
+					 RGWBucketAdminOpState& op_state,
+					 RGWFormatterFlusher& flusher,
+					 const DoutPrefixProvider *dpp,
+					 optional_yield y,
```

Comment:
this function takes optional_yield so there's an expectation that it can run asynchronously, but `io_context::run()` is synchronous. this is fine for radosgw-admin, but we also serve this 'bucket check' operation over the /admin/bucket api

when a yield context is present, we want to use that as the 'parent' of these child coroutines by passing that yield_context as the first argument instead of the io_context. this runs the coroutines on the same thread pool as the frontend requests

in this async case, we can't rely on io_context.run() to block until the children finish, so we need another way to wait asynchronously. when Cory added concurrency to the DeleteObjects op, he combined a counter with an asio timer so the parent coroutine could async_wait() until the number of pending coroutines goes to 0: https://github.com/ceph/ceph/blob/4a05c14f/src/rgw/rgw_op.cc#L6730-L6738

Reply:
As we discussed in Slack, we'll leave this as a future enhancement. Since `check_index_olh` is structured similarly (and was the basis for these changes), it should undergo the same type of enhancement.